### PR TITLE
Feature process

### PR DESCRIPTION
**Pull request by bstansberry**
Beef up the feature process dco; add more metadata to the design doc.

This bases on #581. That branch may change though, in which case I'll need to rebase, so this is currently in draft state. Other than responding to review comments, I've done what I plan though, other than that rebase.

### CODE REVIEW

**Comment by yersan** [comment]
I read the draft and got doubts about how this will collide with the current process for "default" stability level. There is no mention of any subtle distinction for this stability level, so I am not sure if finally, the intention would be whether to merge the requirements described here with the current ones we follow driving features in "default" stability level. 

For example, when we are developing a feature directly to "default" stability level, what sort of rules should we follow? The current rules we follow filling and completing an EAP7? the current rules described in this document? both rules?

Maybe this could be more clear in this draft.

**Comment by bstansberry** [comment]
@yersan This is a WildFly feature process. How EAP chooses to deal with its own internal requirements is not relevant to this document.  If Red Hat asks its employees to do additional things when developing features that will appear in EAP, it's up to Red Hat to document those in its own internal policies.

It's certainly likely that there may be redundant tasks, but this is a pretty lightweight process so I don't think there will be much.

If there's anything in here that conflicts with how you expect things to work when you develop features for inclusion in EAP, please note those things. I, of course, am familiar with EAP's processes and I don't think there is a conflict, but I may have missed something.

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:28]: IMO the wildfly-dev list should be the full address, I am not sure if non subscribers can post so if not maybe we should also point to sign up information here.

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:44]: Is it also worth some comment that we only plan to have iterations for releases N and N+1 so if a feature is a longer running activity they may not be able to set the iteration?

Also do we consider iteration an "exit criteria" for this stage or useful info?

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:60]: I wonder if that could be the time to use the assignee field?  I don't find pings / notifications scale well nowadays but reviewing which issues I am assigned gives me something I can periodically look at..

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:102]: The text below already exists but I am wondering if we should have some clarification in this doc to say things along the lines of, this process should be followed for:

-  The introduction of a feature at a stability level.
- Further feature development at the same stability level.
- AND for the promotion between stability levels.

In all cases requiring an analysis doc covering the relevant progression.

**Review by darranl — APPROVED** [review]
Overall looks good, I have added some comments of things that I thought about as reading but nothing to suggest blocking the merge.

File: FEATURE_PROCESS.adoc
Comment (darranl) [reply FEATURE_PROCESS.adoc:52]: @jamezp what did you think?

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:28]: +1 to the full mail address. I was too lazy to look it up when I drafted this. I'll add the sign up link.

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:120]: I added this; see "Feature Promotion" after "The Process".

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:56]: Signoff?

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:48]: Done.

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc]: Done.

**Comment by bstansberry** [comment]
@darranl You approved this, but I have altered it fairly substantially; mostly for stuff related to your comments but also bits and pieces that occurred to me as I edited. (I think those bits and pieces were primarily adding a bit of rationale for why we want a JIRA and the feature issue, plus various minor copy edits / rendering improvements etc.

**Review by jamezp — COMMENTED** [review]
Sorry, I thought my comment came through, but it was pending. Only one comment from me :)

File: FEATURE_PROCESS.adoc
Comment (yersan) [reply FEATURE_PROCESS.adoc]: A very minor point under your discretion. Could we also add an email subject format here? It could be useful to anyone who wants to filter and classify these message using the email client.

Following what has been sent in the past, the subject could be:

"Feature Team Formation - [stability level] [Jira issue] Short description"

**Comment by darranl** [comment]
I will take another pass through

File: FEATURE_PROCESS.adoc
Comment (darranl) [reply FEATURE_PROCESS.adoc]: If we want something to be in a specific form maybe it could be a part of automation on issue creation to send the initial message?

**Comment by darranl** [comment]
@emmartins I also added you as a reviewer as this talks about adding features to the WildFly quickstarts

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:32]: I think JIRA should be Issue

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:36]: Maybe within the Planning section we can also say something like also not setting the iteration if as a team they are unsure and want to work on the feature first.

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:52]: This is the kind of thing we can add later but as an open source project maybe we could make some suggestions about discussing publicly such as starting a Zulip thread or a thread on the dev list so others with an interest can follow the conversation.

File: FEATURE_PROCESS.adoc
Comment (darranl) [reply FEATURE_PROCESS.adoc:64]: Maybe we need to combine the naming discussion with the previous comment, I think ready for merge would be good so periodically we can look at what is ready and get it merged - but do we want a specific state for "Signoff" / "Reiew" or as that discussion is within the feature team could that be the last thing they do whilst in "In Progress"?

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:105]: I am not sure if this should be phrased differently so it is not about "expertise", I think it is often more a case that the developer is too close to the feature being developed and may not automatically see the view of others coming from a different perspective who may themselves be experts on a different level.  e.g. Experts in the internals of a subsystem vs experts in usability design.

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:108]: This came up in another discussion today, would we want to exclude people from other disciplines here such as professional quality engineers / writers?

File: FEATURE_PROCESS.adoc
Comment (darranl) [inline FEATURE_PROCESS.adoc:117]: Maybe we should add an expectation that if you introduce a feature at the experimental level you will directly engage the community for feedback , i.e. at experimental we are probably looking for a larger representation of "outside perspective" not smaller as this is the time changes can be made.

**Comment by darranl** [comment]
@bstansberry I have taken another pass and my review still stands.

I think apart from the one comment about moving the JIRA issue when I think we really mean GitHub issue the rest is the kinf of thing we incrementally add to and not get caught up discussing every detail.

File: design-doc-template.adoc
Comment (emmartins) [inline design-doc-template.adoc:109]: I think this should be more detailed, perhaps:
`__<Does this feature change the behavior of deployments in incompatible ways? If yes please detail what is the deployment issue observed when no change is done, and what is the change needed to solve the deployment issue>__`

**Review by emmartins — APPROVED** [review]
I did a review from the migration and qs lead perspective, thus abstracting from content not related with those components. I generally approve the changes, but suggest more detail wrt app migration issues

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:64]: I think you get a big vote on that one, as you'd be a prime beneficiary of the info provided by separating 'Review/Signoff' from 'In Progress'.

The other factor is we talked about a transition to 'Review' triggering automation to add a stability-specific checklist to the issue description, a la what Jakarta has in some of their issues. We could do that when the issue moves into In Progress, but then if the team changes the stability target things get messy.

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:32]: Will fix

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:36]: Ok

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:52]: They are already asked to start a wildfly-dev thread. 

But adding text to encourage public discussion is ok. I am concerned though about this turning into a mega document that no one reads.

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc]: If we do anything like this, let's use automation. I'd like to keep this as free of small details as possible so people might actually read it. :)

Also, these emails are good candidates for starting threads for continued discussion of the feature itself and not just the process.

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:56]: In a discussion today we agreed to drop the 'Review/Signoff' content until their is some sort of automation in place to support it, i.e. adding a stability-specific checklist to the issue description that the feature team would check off. And it would probably make more sense to do that addition when the issue moves to In Progress.

We also agreed to add a 'Ready for Merge' status, to which the feature team should move the issue once all requirements have been met. We discussed some automation around that (i.e. team checks off requirements and automation moves it when done), but that's for the future.

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:64]: See above; for now we decided to drop the 'Review/Signoff' stage unless their is automation to bring value to it.

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:36]: Done

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:52]: Added.

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:108]: Not at all. For 'default' stability features my assumption is the quality engineers and writers we want involved would often act in this role, particularly the writers.

Comment (bstansberry) [reply FEATURE_PROCESS.adoc:108]: Also, I don't want to get into this detail in this document, but as a practical matter if the team includes a professional QE person and a technical writer, the team has an 'Outside Perspective'. It seems like a real corner case where that wouldn't be reasonably true.

File: FEATURE_PROCESS.adoc
Comment (bstansberry) [reply FEATURE_PROCESS.adoc:117]: I added a sentence.

File: design-doc-template.adoc
Comment (bstansberry) [reply design-doc-template.adoc:109]: updated; thanks

**Pull request merged**
